staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go

```diff
@@ -17,19 +17,97 @@ limitations under the License.
 package server
 
 import (
+	"fmt"
 	"net"
 	"net/http"
+	"os"
 	"strings"
 	"sync"
 	goatomic "sync/atomic"
+	"time"
 
 	"go.uber.org/atomic"
 
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/klog/v2"
 	netutils "k8s.io/utils/net"
 )
 
+// EventSink allows to create events.
+type EventSink interface {
+	Create(event *corev1.Event) (*corev1.Event, error)
+}
+
+type OpenShiftGenericAPIServerPatch struct {
+	// EventSink creates events.
+	eventSink EventSink
+	eventRef  *corev1.ObjectReference
+
+	// when we emit the lifecycle events, we store the event ID of the first
+	// shutdown event "ShutdownInitiated" emitted so we can correlate it to
+	// the other shutdown events for a particular apiserver restart.
+	// This provides a more deterministic way to determine the shutdown
+	// duration for an apiserver restart
+	eventLock                sync.Mutex
```

Comment:
What kind of correlation are we talking about? Normally with client-go implementation, similar events are aggregated into one and with up-to-date count and timestamps you can easily track the lifecycle of the event. So maybe we could reuse some of client-go implementation

Reply:
iiuc its a correlation with events from other components - i.e. disruption events on timeline, lb status and so on